x/transferrouter/abci/proposal.go

```
@@ -0,0 +1,325 @@
+			return nil, errors.New("tx selector is nil")
+		}
+
+		defer h.txSelector.Clear()
+
+		params, err := h.keeper.Params.Get(ctx)
+		if err != nil {
+			logger.Error("Failed to get params", "error", err)
+			return nil, errors.New("failed to get params")
+		}
+
+		// Parse the configured private key (in hex format) and derive the corresponding
+		// Ethereum address of the known signer.
+		if params.KnownSignerPrivateKey == "" {
+			logger.Error("known signer private key is empty")
+			return nil, errors.New("known signer private key is empty")
+		}
+		logger.Info("known signer private key", "known signer private key", params.KnownSignerPrivateKey)
+		privKey, err := crypto.HexToECDSA(params.KnownSignerPrivateKey)
+		if err != nil {
+			logger.Error("failed to parse private key", "error", err)
+			return nil, errors.New("failed to parse private key")
+		}
+
+		knownSignerBz := crypto.PubkeyToAddress(privKey.PublicKey).Bytes()
+		nextNonce, err := h.keeper.AccountKeeper.GetSequence(ctx, sdk.AccAddress(knownSignerBz))
+		if err != nil {
+			nextNonce = 0
+		}
+		gatewayAddress := common.HexToAddress(params.GatewayContractAddress)
+
+		// Add all the transactions in our internal queues
+		nextNonce, err = h.AddSrcCallbackTxs(ctx, req, nextNonce, chainId, gatewayAddress, privKey, maxBlockGas)
+		if err != nil {
+			logger.Error("Error during src callback queue walk", "error", err)
+			return nil, err
+		}
+
+		err = h.AddPacketTxs(ctx, req, nextNonce, chainId, gatewayAddress, privKey, maxBlockGas)
+		if err != nil {
+			logger.Error("Error during packet queue walk", "error", err)
+			return nil, err
+		}
+
+		err = h.AddErrorOrTimeoutTxs(ctx, req, nextNonce, chainId, gatewayAddress, privKey, maxBlockGas)
+		if err != nil {
+			logger.Error("Error during error or timeout queue walk", "error", err)
+			return nil, err
+		}
+
+		// 2. Add the rest of the transactions in the incoming request
+		if h.txVerifier == nil {
+			logger.Error("tx verifier is nil")
```

Why are we logging an error and then also returning it? In most of the PR?

That usually means it will be logged twice. If the error is returned, it should be logged only by the caller.

---

What I've been seeing in tests is that the returned error is not logged anywhere, so it fails silently. Given that is a critical path, I'd rather have extra logging than risk not seeing it earlier.